### PR TITLE
Add Prewarming Pool Alert

### LINKED ISSUE #470: Lost Runners, again
On October 4th, we lost some runners again, leading to an empty pre-warming pool and errors being displayed to our learners.

Timestamps shown in CEST:
Bildschirmfoto 2023-10-04 um 08 38 48

### PR DESCRIPTION
**Pull request by mpass99**
Closes #470 

- that checks for every environment if the filled share of the prewarming pool is at least the specified threshold.
- ToDo

### CODE REVIEW

File: internal/api/health.go [hunk 1]
```diff
@@ -1,12 +1,42 @@
...
+
 // Health handles the health route.
 // It responds that the server is alive.
 // If it is not, the response won't reach the client.
-func Health(writer http.ResponseWriter, _ *http.Request) {
-	writer.WriteHeader(http.StatusNoContent)
+func Health(manager environment.Manager) http.HandlerFunc {
+	return func(writer http.ResponseWriter, request *http.Request) {
+		if err := checkPrewarmingPool(manager); err != nil {
+			writeInternalServerError(writer, err, dto.ErrorUnknown, request.Context())
```
Comment (MrSerth) [inline internal/api/health.go:20]: We could also introduce a new error code here; it's not that "unknown", is it?

File: internal/api/health.go [hunk 2]
```diff
@@ -1,12 +1,42 @@
...
+	}
+}
+
+func checkPrewarmingPool(manager environment.Manager) error {
+	prewarmingPoolOk := true
+	err := ErrorPrewarmingPoolDepleting
+	for _, data := range manager.Statistics() {
+		if float64(data.IdleRunners)/float64(data.PrewarmingPoolSize) < config.Config.Server.Alert.PrewarmingPoolThreshold {
+			prewarmingPoolOk = false
+			err = fmt.Errorf("%w: environment %d", err, data.ID)
```
Comment (MrSerth) [inline internal/api/health.go:34]: This can lead to some interesting error messages:

```
    "message": "the prewarming pool is depleting: environment 29: environment 34",
```

Here, I would have expected that only one colon (`:`) is included and the second environment is added with a comma (`,`).

File: internal/runner/nomad_manager.go [hunk 3]
```diff
@@ -52,6 +55,44 @@ func (m *NomadRunnerManager) Claim(environmentID dto.EnvironmentID, duration int
...
+	defer mutex.Unlock()
+
+	prewarmingPoolThreshold := config.Config.Server.Alert.PrewarmingPoolThreshold
+	reloadTimeout := config.Config.Server.Alert.PrewarmingPoolReloadTimeout
+
+	if reloadTimeout == 0 || float64(environment.IdleRunnerCount())/float64(environment.PrewarmingPoolSize()) >= prewarmingPoolThreshold {
+		return
+	}
+
+	log.WithField(dto.KeyEnvironmentID, environment.ID()).Info("Prewarming Pool Alert. Checking again..")
```
Comment (MrSerth) [inline internal/runner/nomad_manager.go:75]: Even though I configured `prewarmingpoolreloadtimeout: 30`, I never get this line (and no reloading):

```log
time="2023-11-04T18:10:30.069365Z" level=debug code=204 duration="40µs" method=GET path=/api/v1/health user_agent=PostmanRuntime/7.34.0
[Sentry] 2023/11/04 18:10:30 Event dropped due to noopTransport usage.
time="2023-11-04T18:10:38.061950Z" level=debug msg="Handle Allocation Event" ClientStatus=running DesiredStatus=stop NextAllocation= PrevAllocation= alloc_id=c0caa42d-4cf8-5d48-2d4d-dddd694d531c package=nomad
time="2023-11-04T18:10:38.399774Z" level=debug msg="Handle Allocation Event" ClientStatus=complete DesiredStatus=stop NextAllocation= PrevAllocation= alloc_id=c0caa42d-4cf8-5d48-2d4d-dddd694d531c package=nomad
time="2023-11-04T18:10:38.399895Z" level=debug msg="Runner stopped" package=runner runner_id=34-0a49bd1c-7b35-11ee-836b-2e62a1fdfe46
[Sentry] 2023/11/04 18:10:51 Event dropped due to noopTransport usage.
time="2023-11-04T18:10:51.797806Z" level=error msg="Failing /api/v1/health" code=500 duration="168µs" method=GET path=/api/v1/health user_agent=PostmanRuntime/7.34.0
[Sentry] 2023/11/04 18:10:51 Event dropped due to noopTransport usage.
time="2023-11-04T18:12:01.034826Z" level=debug msg="Handle Allocation Event" ClientStatus=running DesiredStatus=stop NextAllocation= PrevAllocation= alloc_id=4b88183f-581b-3d2d-d7a1-eeafd374af08 package=nomad
time="2023-11-04T18:12:01.204157Z" level=debug msg="Handle Allocation Event" ClientStatus=complete DesiredStatus=stop NextAllocation= PrevAllocation= alloc_id=4b88183f-581b-3d2d-d7a1-eeafd374af08 package=nomad
time="2023-11-04T18:12:01.204246Z" level=debug msg="Runner stopped" package=runner runner_id=29-0a435ee0-7b35-11ee-836b-2e62a1fdfe46
[Sentry] 2023/11/04 18:12:03 Event dropped due to noopTransport usage.
time="2023-11-04T18:12:03.395046Z" level=error msg="Failing /api/v1/health" code=500 duration="71µs" method=GET path=/api/v1/health user_agent=PostmanRuntime/7.34.0
[Sentry] 2023/11/04 18:12:03 Event dropped due to noopTransport usage.
[Sentry] 2023/11/04 18:17:53 Event dropped due to noopTransport usage.
time="2023-11-04T18:17:53.293006Z" level=error msg="Failing /api/v1/health" code=500 duration="23µs" method=GET path=/api/v1/health user_agent=PostmanRuntime/7.34.0
[Sentry] 2023/11/04 18:17:53 Event dropped due to noopTransport usage.
```

In order to test the system, I was just synching an environment through CodeOcean, and was then using the Nomad UI to stop (and purge) the allocations. For the health route, this worked as expected :)

File: internal/runner/nomad_manager.go [hunk 4]
```diff
@@ -95,69 +134,100 @@ func (m *NomadRunnerManager) SynchronizeRunners(ctx context.Context) error {
...
+		newJob.SetupTimeout(time.Duration(timeout) * time.Second)
 	} else {
 		environment.AddRunner(newJob)
 	}
+	return newJob, isUsed, nil
 }
 
-func (m *NomadRunnerManager) updateUsedRunners(newUsedRunners storage.Storage[Runner]) {
+// updateUsedRunners handles the update of the used runners cleanly. This means it removes the runners not in newUsedRunners.
+// Iff deleteOthers all runners will not in newUsedRunners will be destroyed locally. Otherwise, just the replaced runners.
```
Comment (MrSerth) [inline internal/runner/nomad_manager.go:216]: This comment is hard to read, could you rephrase it, please?

File: api/swagger.yaml [hunk 5]
```diff
@@ -162,7 +162,8 @@ paths:
       responses:
         "204":
           description: Everything okay
-
+        "500":
```
[the remark this answers is not in the document]
Comment (MrSerth) [reply api/swagger.yaml:165]: We're quite lucky that we named that route `/health` 😅.

While the answer given in StackOverflow is correct, handling a non-2xx code is much easier in the different monitoring tools than parsing the error further. Therefore, I still have some (slight) preference in the direction of using a 5xx code, but we can discuss that.

File: internal/runner/nomad_manager.go (hunk 3 above)
Comment (mpass99) [reply internal/runner/nomad_manager.go:75]: That's the case because the check for the health route is triggered on each request to the `health` route and
the prewarming pool reloading is triggered on each request to the `provide` runner route.
We could also easily trigger the check on each request to the health route. However, this has its ups and downsides. On the one hand, currently, we would only resync those environments being used. On the other hand, when we check the prewarming pool on every health request, we could fix prewarming pool issues when they appear and not wait until someone is annoyed from receiving an error message.

File: internal/runner/nomad_manager.go [hunk 6]
```diff
@@ -95,69 +134,100 @@ func (m *NomadRunnerManager) SynchronizeRunners(ctx context.Context) error {
...
+		for _, r := range usedRunners.List() {
+			newUsedRunners.Add(r.ID(), r)
 		}
-		err = environment.ApplyPrewarmingPoolSize()
+	}
+
+	m.updateUsedRunners(newUsedRunners, true)
+}
+
+func (m *NomadRunnerManager) loadEnvironment(environment ExecutionEnvironment) (used storage.Storage[Runner], err error) {
```
Comment (mpass99) [inline internal/runner/nomad_manager.go:165]: What is the difference between `NomadRunnerManager.loadEnvironment` and `NomadEnvironment.ApplyPrewarmingPoolSize`? Are we creating parallel structures for the same task? Shouldn't it be in the control of the environment to manage its runners?

I did so as the Nomad Environment...
1. has no control over the used runners (which otherwise would not be loaded).
2. cannot create Nomad Runner References (just indirectly by requesting a new runner at Nomad and Nomad notifying the Runner Manager via the Event Stream).
3. the load/recover functionality was historically always directly in the runner manager.

Even though it's not optimal, I would go with it. What do you think?

File: internal/runner/nomad_manager.go (hunk 3 above)
Comment (MrSerth) [reply internal/runner/nomad_manager.go:75]: Ah, okay! I didn't consider that the `provide` route is being used there. From the documentation, I was (wrongly) assuming that there is a direct connection to the `health` route:

```yaml
    # The prewarming pool reload timeout (in seconds) defines for how long the low prewarming pool warning (above)
    # should be active before Poseidon automatically reloads the environment.
```

This sentence made that connection for me; therefore I am a little surprised about that (but get it now). For sure, having the reloading on the `provide` route makes sense (I could image a situation where Poseidon's `health` route is not regularly called -- the second mechanism should still work). Therefore, I would argue not to add it only to the `health` route. However, one could also argue that the whole alert-thing should be somewhat de-coupled from any route (the state change to warning also happens without a request, right?). Not sure, though. Let's shortly discuss the pros and cos later.

File: internal/runner/nomad_manager.go (hunk 6 above)
Comment (MrSerth) [reply internal/runner/nomad_manager.go:165]: Actually, I don't see where we are duplicating code or behaviour in these two methods. Sure, `NomadRunnerManager.loadEnvironment` is calling `NomadEnvironment.ApplyPrewarmingPoolSize`, but that feels right. Otherwise, and especially because `NomadEnvironment.ApplyPrewarmingPoolSize` is so small, I don't see an issue and would proceed as you suggested.

File: internal/runner/nomad_manager.go (hunk 4 above)
Comment (MrSerth) [reply internal/runner/nomad_manager.go]: Thanks, much better (now I get it 😁)

**Comment by MrSerth** [comment]
Awesome, this really looks nice and I am happy with the refactoring that you performed. There is just one question that arose when checking the logs. Consider the following output:

```
time="2023-11-09T09:17:53.973088Z" level=debug msg="Runner started" package=runner runner_id=template-34 startupDuration=590.470375ms
time="2023-11-09T09:17:53.974021Z" level=debug msg="Handle Allocation Event" ClientStatus=running DesiredStatus=run NextAllocation= PrevAllocation= alloc_id=28e9a082-06da-9f47-df18-13938b161467 package=nomad
time="2023-11-09T09:17:53.974057Z" level=debug msg="Runner started" package=runner runner_id=34-7ad43840-7ed8-11ee-9965-2e62a1fdfe45 startupDuration=584.227708ms
time="2023-11-09T09:17:53.975797Z" level=debug msg="Handle Allocation Event" ClientStatus=running DesiredStatus=run NextAllocation= PrevAllocation= alloc_id=abf5b573-b045-43fa-95cf-cccdb639e45b package=nomad
time="2023-11-09T09:17:53.975854Z" level=debug msg="Runner started" package=runner runner_id=34-7ad54fbe-7ed8-11ee-9965-2e62a1fdfe45 startupDuration=580.006125ms
time="2023-11-09T09:18:13.790508Z" level=debug msg="Handle Allocation Event" ClientStatus=running DesiredStatus=stop NextAllocation= PrevAllocation= alloc_id=28e9a082-06da-9f47-df18-13938b161467 package=nomad
time="2023-11-09T09:18:14.082956Z" level=debug msg="Handle Allocation Event" ClientStatus=complete DesiredStatus=stop NextAllocation= PrevAllocation= alloc_id=28e9a082-06da-9f47-df18-13938b161467 package=nomad
time="2023-11-09T09:18:14.083141Z" level=debug msg="Runner stopped" package=runner runner_id=34-7ad43840-7ed8-11ee-9965-2e62a1fdfe45
time="2023-11-09T09:18:14.083261Z" level=info msg="Prewarming Pool Alert. Checking again.." environment_id=34 package=runner
time="2023-11-09T09:18:44.083890Z" level=warning msg="Prewarming Pool Alert. Reloading environment" environment_id=34 package=runner
time="2023-11-09T09:18:44.087935Z" level=debug msg="Recovered Allocation" jobID=34-7ad54fbe-7ed8-11ee-9965-2e62a1fdfe45 package=nomad status=running
time="2023-11-09T09:18:44.095918Z" level=debug msg="Recovered Runner" isUsed=false package=runner runner_id=34-7ad54fbe-7ed8-11ee-9965-2e62a1fdfe45
time="2023-11-09T09:18:44.096394Z" level=debug msg="Destroying Runner" destroy_reason="the runner will be destroyed and replaced: the destruction should not cause external changes" package=runner runner_id=34-7ad54fbe-7ed8-11ee-9965-2e62a1fdfe45
time="2023-11-09T09:18:44.096426Z" level=debug msg="Runner destroyed locally" destroy_reason="the runner will be destroyed and replaced: the destruction should not cause external changes" package=runner runner_id=34-7ad54fbe-7ed8-11ee-9965-2e62a1fdfe45
time="2023-11-09T09:18:44.096448Z" level=debug msg="Creating new runners" environment_id=34 package=environment runnersRequired=1
time="2023-11-09T09:18:44.103942Z" level=debug msg="Handle Allocation Event" ClientStatus=pending DesiredStatus=run NextAllocation= PrevAllocation= alloc_id=1da620b0-ed86-4c0d-9d3e-882401d20d37 package=nomad
time="2023-11-09T09:18:44.527221Z" level=debug msg="Handle Allocation Event" ClientStatus=running DesiredStatus=run NextAllocation= PrevAllocation= alloc_id=1da620b0-ed86-4c0d-9d3e-882401d20d37 package=nomad
time="2023-11-09T09:18:44.527277Z" level=debug msg="Runner started" package=runner runner_id=34-990e81e4-7ed8-11ee-9965-2e62a1fdfe45 startupDuration=423.303458ms
```

Do you know why the `Destroying Runner` and `Runner destroyed locally` messages were sent? As you can see, I stopped runner 34-7ad43840-7ed8-11ee-9965-2e62a1fdfe45, leaving 34-7ad54fbe-7ed8-11ee-9965-2e62a1fdfe45. This runner also got recovered, as written in the log. This runner didn't get changed on Nomad, so was still running after recovering (hence, the destruction indeed didn't cause and external change).

File: internal/runner/nomad_manager.go [hunk 7]
```diff
@@ -95,69 +100,158 @@ func (m *NomadRunnerManager) SynchronizeRunners(ctx context.Context) error {
...
+	m.reloadingEnvironment.Add(environment.ID().ToString(), &alertData{Mutex: &sync.Mutex{}, cancel: nil})
+}
+
+func (m *NomadRunnerManager) DeleteEnvironment(id dto.EnvironmentID) {
+	m.AbstractManager.DeleteEnvironment(id)
+	m.reloadingEnvironment.Delete(id.ToString())
+}
+
+// checkPrewarmingPoolAlert checks if the prewarming pool contains enough idle runners as specified by the PrewarmingPoolThreshold
+// if not it starts an environment reload mechanism according to the PrewarmingPoolReloadTimeout.
```
Comment (MrSerth) [inline internal/runner/nomad_manager.go:114]:
```suggestion
// checkPrewarmingPoolAlert checks if the prewarming pool contains enough idle runners as specified by the PrewarmingPoolThreshold.
// If not, it starts an environment reload mechanism according to the PrewarmingPoolReloadTimeout.
```

File: internal/runner/nomad_manager.go [hunk 8]
```diff
@@ -95,69 +100,158 @@ func (m *NomadRunnerManager) SynchronizeRunners(ctx context.Context) error {
...
+	if runnerAdded && data.cancel != nil {
+		data.cancel()
+		data.cancel = nil
+		m.checkPrewarmingPoolAlert(environment, false)
+		return
+	}
+
+	// With this hard lock we collect/block goroutines waiting for one reload to be done.
+	// However, in practice its likely that only up to PrewarmingPoolSize/2 goroutines are waiting.
+	// We could avoid the waiting, but we use it to solve the race conditions of the recursive call above.
```
Comment (MrSerth) [inline internal/runner/nomad_manager.go:131]:
```suggestion
	// With this hard lock, we collect/block goroutines waiting for one reload to be done.
	// However, in practice, it's likely that only up to PrewarmingPoolSize/2 goroutines are waiting.
	// We could avoid the waiting, but we use it to solve the race conditions of the recursive call above.
```

File: internal/runner/nomad_manager.go [hunk 9]
```diff
@@ -95,69 +100,158 @@ func (m *NomadRunnerManager) SynchronizeRunners(ctx context.Context) error {
...
+	reloadTimeout := config.Config.Server.Alert.PrewarmingPoolReloadTimeout
+
+	if reloadTimeout == 0 || float64(environment.IdleRunnerCount())/float64(environment.PrewarmingPoolSize()) >= prewarmingPoolThreshold {
+		return
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	data.cancel = cancel
+
+	log.WithField(dto.KeyEnvironmentID, environment.ID()).Info("Prewarming Pool Alert. Checking again..")
```
Comment (MrSerth) [inline internal/runner/nomad_manager.go:145]: So far, we mostly used three dots:

```suggestion
	log.WithField(dto.KeyEnvironmentID, environment.ID()).Info("Prewarming Pool Alert. Checking again...")
```

When checking the code, we sometimes have a space before the three dots, sometimes not. Should we unify that (just in a dedicated commit, within this PR)?

File: internal/api/health.go [hunk 10]
```diff
@@ -1,12 +1,42 @@
...
+
+func checkPrewarmingPool(manager environment.Manager) error {
+	var depletingEnvironments []int
+	for _, data := range manager.Statistics() {
+		if float64(data.IdleRunners)/float64(data.PrewarmingPoolSize) < config.Config.Server.Alert.PrewarmingPoolThreshold {
+			depletingEnvironments = append(depletingEnvironments, data.ID)
+		}
+	}
+	if len(depletingEnvironments) > 0 {
+		arrayToString := strings.Trim(strings.Join(strings.Fields(fmt.Sprint(depletingEnvironments)), ","), "[]")
```
Comment (MrSerth) [inline internal/api/health.go:38]:
```suggestion
		arrayToString := strings.Trim(strings.Join(strings.Fields(fmt.Sprint(depletingEnvironments)), ", "), "[]")
```

**Review by MrSerth — COMMENTED** [review]
Here are just some minor text suggestions for the comments / strings.

**Comment by mpass99** [comment]
> Do you know why the `Destroying Runner` and `Runner destroyed locally` messages were sent?

These messages were sent because we log for debug purposes each time before and after a runner gets destroyed and why.

This runner is not destroyed at Nomad, but just the reference in Poseidon. This is the case because we perform a full recovery. We first create new runners (including maybe changed values like the portMapping or if it is used). Then, we replace the old runner with the new one.

One advantage of doing so is to keep the interface of the Nomad runner simple. Simple, in the recovery process and to spare methods for changing i.e. the portMapping.
One disadvantage is that we currently lose (/don't care about) the executions of that runner. Meaning, that in the (small) time between the creation of an execution and the start of the execution, a recovery would lead to an error. But, as we only rarely recover (expecting Nomad/the environment is having problems anyway), we may accept this risk.

**Comment by MrSerth** [comment]
Alright, that makes sense -- thank you!

**Review by MrSerth — APPROVED** [review]
